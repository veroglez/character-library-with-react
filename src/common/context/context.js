import { node } from 'prop-types';
import React, { createContext, Component } from 'react';
import { C, L10N } from '..';

const Context = createContext();
const { Provider, Consumer: ConsumerData } = Context;

class ProviderData extends Component {
  static propTypes = {
    children: node,
  };

  static defaultProps = {
    children: undefined,
  };

  state = {
    next: '',
    characters: [],
    l10n: L10N,
  }

  componentDidMount() {
    const service = 'character';
    this.onFetch(service);
  }

  onFetch = async (service, next) => {
    const { characters } = this.state;
    const response = await fetch(next || `${C.DOMAIN}${service}`);
    const character = await response.json();
    this.setState({ characters: next ? [...characters, ...character.results] : character.results, next: character.info.next });
  }

  onData = object => this.setState(object);

  render() {
    const {
      onData, onFetch, props: { children },
    } = this;

    return (
      <Provider value={{ ...this.state, onData, onFetch }}>
        { children }
      </Provider>
    );
  }
}

export { ConsumerData, ProviderData };
