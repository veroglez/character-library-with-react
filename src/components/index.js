import Card from './Card';
import CardList from './CardList';
import Detail from './Detail';
import Header from './Header';
import Image from './Image';
import SearchInput from './InputSearch';
import Text from './Text';
import Tags from './Tags';

export {
  Card,
  CardList,
  Detail,
  Image,
  Header,
  SearchInput,
  Tags,
  Text,
};
