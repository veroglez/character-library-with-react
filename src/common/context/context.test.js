import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { ConsumerData, ProviderData } from './context';
import { Text } from '../../components';

Enzyme.configure({ adapter: new Adapter() });

describe('<Context>', () => {
  it('<ProviderData> renders', () => {
    const tree = renderer.create(<ProviderData />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('<ConsumerData> renders', () => {
    const tree = renderer.create(
      <ConsumerData>
        { () => <Text body text="test" /> }
      </ConsumerData>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('<Context> renders', () => {
    const tree = renderer.create(
      <ProviderData>
        <ConsumerData>
          { context => Object.keys(context) }
        </ConsumerData>
      </ProviderData>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('<ConsumerData> can get properties of context', () => {
    const tree = renderer.create(
      <ProviderData>
        <ConsumerData>
          { ({ characters }) => <Text body text={characters} /> }
        </ConsumerData>
      </ProviderData>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('onData method', () => {
    const wrapper = shallow(<ProviderData />);

    wrapper.instance().onData({ test: 'I am doing a test' });
    expect(wrapper.instance().state.test).toEqual('I am doing a test');
  });
});
