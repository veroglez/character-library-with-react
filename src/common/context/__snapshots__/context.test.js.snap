// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Context> <ConsumerData> can get properties of context 1`] = `
<p
  className="undefined undefined"
/>
`;

exports[`<Context> <ConsumerData> renders 1`] = `
<p
  className="undefined undefined"
/>
`;

exports[`<Context> <Context> renders 1`] = `
Array [
  "next",
  "characters",
  "l10n",
  "favourites",
  "onData",
  "onFetch",
]
`;

exports[`<Context> <ProviderData> renders 1`] = `null`;
